<h1 class="r-stretch" style="text-align:center"> Introduction to Python, Part I  </h1>

<center>
    <br>
     <table style="background-color: white;">
        <tr>
            <td>
    <img src="https://www.python.org/static/community_logos/python-logo.png" width=250 style="float:center">
            </td>
            <td>
    <img src="https://brand.jhu.edu/assets/uploads/sites/5/2014/06/university.logo_.small_.horizontal.blue_.jpg" width=250 style="float:center">
            </td>
            </tr>
    </table>
        <br>
    Johns Hopkins University  <br>
Krieger School of Arts and Sciences  <br>
Office of Undergraduate Research, Scholarly and Creative Activity (URSCA) <br>
May 23, 2022 <br>
    <br>
    Prof. Daniel Beller <br>
    JHU Department of Physics &amp; Astronomy
    </center>
 </section>

<div style="float:right">
    <img src="https://www.python.org/static/community_logos/python-logo.png" width=250>
</div>

## What is Python?  

<br>

<table width=100%>
    <tr>
        <td>
Python is a <span class="new_term">computer programming language</span>: <br><span class="fragment"> <ul><li> rules for turning a text document written by a human (the <span class="new_term"> program file</span>) into instructions that a computer can run (the <span class="new_term">executable</span>).</li></ul></span>   <br> <br> 
    <span class="fragment">Compared to other programming languages, Python was designed to be more similar to English and easier to read.</span>
        </td>
        <td width=40%>
            <table>
                <tr>
                    <td>
            <img src="https://victoria.mediaplanet.com/app/uploads/sites/143/2020/03/Female-data-scientist-at-work-888x500.jpg" width=250> 
                    </td>
                    <td>→
                    </td>
                    <td>
            
<img src="https://cdn3.iconfinder.com/data/icons/data-sharing-flat-colorful/614/4963_-_Computer_Binary_Code-512.png" width=250>
                    </td>
            </table>
                </td>
    </tr>
    </table>



## Why learn Python?

<div height=80%>
        <div class = "fragment fade-in-then-semi-out">
        <h4> Useful in a wide range of careers </h4>
    Python has become one of the most popular programming languages for:
        <ul>
            <li> scientific computing </li>
            <li> data science ("big data") </li>
            <li> machine learning (artificial intelligence) </li>
            <li> computerized tasks in a variety of industries such <br> as web development and finance </li>
        </ul>
    <img src="https://myseco.militaryonesource.mil/Portal/Media/Default/O-Images/businesswomen-discussing-coding-on-computer-g1180183927.jpg?content" align="right" width=25% style="  border: 2px solid #111;">
    </div>
     <div class="fragment fade-in-then-semi-out	">
         <h4> It's free </h4> 
         Unlike some popular alternatives such as Matlab.
    </div>
    <div class="fragment fade-in-then-semi-out	">
        <h4> Python is a great language for <em>scripting</em> </h4>
        <img src="https://automatetheboringstuff.com/images/cover_automate2_thumb.jpg" style="float:right" width=15%>
           automating small tasks with text, math, and web data        
    </div>
    <div class="fragment fade-in-then-semi-out	">
        <h4> Huge library of free software </h4>
        from astrophysics to writing music to running <em> Minecraft </em>
        <img src="https://pypi.org/static/images/logo-large.6bdbb439.svg" style="float:right">
    </div>
    <div class="fragment fade-in-then-semi-out">
        <h4>Gateway to programming </h4> 
            <table style="float:right;" width=25%>
                <tr>
                    <td>
            <img src="https://upload.wikimedia.org/wikipedia/commons/1/19/C_Logo.png">
                    </td>
                    <td>
            <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/1/18/ISO_C%2B%2B_Logo.svg/306px-ISO_C%2B%2B_Logo.svg.png">
                    </td>        
                    <td>
            <img src="https://logoeps.com/wp-content/uploads/2013/03/java-eps-vector-logo.png">
                    </td>
                </tr>
        </table>
        Python is pretty similar to other widely used programming languages like C, C++, and Java. <br>
        Once you learn some Python, you have a big head start in learning other languages.             <br><br>
    </div>
</div>



## Goals for these sessions

* Learn the fundamentals of the Python language

* Write and run your own program

* You'll gain the necessary foundation to teach yourself more, if you'd like 
<img class="fragment fade-up" src="tutorial-screenshot.png" style="float:right" width=40%>

## How can I use Python?

We need a **Python interpreter**: Special program to translate our program file into an executable. 

* Usually when we say "Python" we mean both the language and the intepreter. 

**The traditional way:** Download the interpreter from www.python.org and install. 

**What we'll do today:** Use a web-based interpreter. 



1. In web browser, go to [jupyter.org/try](https://jupyter.org/try).
1. Scroll down and click on the Python symbol <img src="https://jupyter.org/assets/try/python.svg" width=15%>
1. Go File -> New -> Notebook 
1. Press `OK` to accept "Pyolite" as the kernel 

<img src="try-screenshot.png" width=40% style="float:right">
<br>
<br>

## Writing your first Python program

Click on the empty gray bar to start typing. Enter this:

In [ ]:
print("Hello world")

Then hold down `Shift` and press `Enter`. You should see:

In [35]:
print("Hello world")

Hello world


That's it! You've written your first Python program! 🤗

## Math basics

The Python interpreter can work as a calculator. Scientists who use Python rely on this extensively.

### Integers

Addition

In [36]:
1 + 2

3

Subtraction

In [234]:
2 - 8

-6

### Integers (continued)

Use asterisk `*` for multiplication.

In [2]:
3 * 5

15

Use `//` to divide one integer by another.

In [39]:
16 // 2

8

Use `%` for remainder

In [40]:
7 // 2

3

In [41]:
7 % 2

1

Exponents use `**`. 

For example, to calculate $3^2$:

In [43]:
3 ** 2 

9

Python math obeys order of operations: 

In [44]:
4 + 3 * 2**2

16

Use parentheses `()` to change order of operations

In [45]:
(4 + 3) * 2**2 

28

## Math basics: floats

A number with a decimal point is called a <span class="new_term">float</span>. 

Python math works mostly the same with floats as with integers. 

Let's divide a $76.15 dinner bill among four friends:

In [46]:
76.15 / 4

19.0375

In real life we'd round to the nearest cent: $19.04. Python can even help us do this:

In [47]:
round(76.15 / 4, 2)

19.04

What's 71% of 505, out to 2 decimal places?

In [9]:
round(0.71 * 505, 2)

358.55

<span class="your_turn"> Try it yourself </span>

Use Python to calculate an **18%** tip for your waiter if your dinner bill was **$19.04**.

In [48]:
round(19.04 * 0.18, 2)

3.43

The tip is $3.43.

## Comments

If you type `#`, everything after that on the same line is a <span class="new_term">comment</span>.   

A comment is just for human programmers to read. The computer ignores it.

In [2]:
# This is a comment! 

2.5 * 2 # In this line we multiply 2.5 by 2

5.0

Use comments to explain your program in plain English.

In [ ]:
(4 + 3) * 2**2 # use parentheses so addition happens first

<span class="your_turn"> Try it yourself </span> — add a comment to your Hello World program.


## Errors

If any line in your program doesn't make sense to the Python interpreter, the program will stop and show you an <span class="new_term">Error</span> message.

In [68]:
my_first_variable = 3.5 
my second variable = 4.5

SyntaxError: invalid syntax (<ipython-input-68-c15316683c82>, line 2)

Don't be embarassed. These "bugs" happen all the time to computer programmers. Python tries to help you "debug". 

<span class="your_turn">Can you see how to fix the error?</span>


<code > my_first_variable = 3.5 </code><br>
<code class="fragment highlight-blue"> my_second_variable = 4.5 </code>
<br><br><br>

## Variables and Assignment

Using the `=` character, we can <span class="new_term">assign</span> a value, such as `5`, to a variable name, such as `x`:

In [50]:
x = 5 # assignment 

In [51]:
x     # show the value assigned to x

5

In [53]:
x + 11 # do math with the variable in place of 5

16

<h3 class="fragment semi-fade-out"> Variables and Assignment </h3>

<span class="fragment">
    By combining <span class="new_term"> assignment </span> with Python math, we can break up a many-step calculation into easy chunks.
    </span>

In [55]:
# The dinner bill for all four friends:
bill = 76.15 

# The share of the bill owed by each friend:
my_bill = bill / 4

# Each friend's 18% tip:
my_tip = my_bill * 0.18 

# How much each friend should pay:
my_total = my_bill + my_tip

# Round to the nearest cent:
my_total_rounded = round(my_total, 2)

# Show how much I should pay
print(my_total_rounded)

22.46


<br><br><br>

## Strings

Besides working with numbers, Python also makes it easy to work with **text**. 

Any chunk of text surrounded by `' '` or `" "` is called a <span class="new_term">string</span>.

In [56]:
print("Python is a great computer programming language")

Python is a great computer programming language


A string can be assigned as the value of a variable.

In [14]:
wonderful_phrase = "Hakuna Matata"

print(wonderful_phrase)

Hakuna Matata


<h2 class="fragment fade-in-then-semi-out">Strings</h2>

Use the `in` keyword  to check if a string contains a word (or letter, or phrase, etc.)

In [15]:
"Matata" in wonderful_phrase # check whether the string contains "Matata"

True

In [16]:
"q" in wonderful_phrase # check whether the letter q appears in the string

False

We can "add" and "multiply" strings:

In [17]:
wonderful_phrase + " Ain't no passing craze"

"Hakuna Matata Ain't no passing craze"

In [18]:
(wonderful_phrase + " ") * 3

'Hakuna Matata Hakuna Matata Hakuna Matata '

Use the `len()` function to find out how many characters are in our string (including the space!)

In [20]:
len(wonderful_phrase)

13

<br><br><br>

If your string takes up multiple lines or has quotation marks in it, surround it by triple-quotes `"""   """` to be safe.

Here are the lyrics to an early Beatles song I like.

In [235]:
lyrics = """
Oh yeah, I'll tell you something
I think you'll understand
When I say that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand

Oh please, say to me
You'll let me be your man
And please, say to me
You'll let me hold your hand
You'll let me hold your hand
I wanna hold your hand

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I say that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I feel that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand
"""

<br><br><br>


<h2 class="fragment semi-fade-out"> Strings (continued) </h2>

In [151]:
print(lyrics)


Oh yeah, I'll tell you something
I think you'll understand
When I say that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand

Oh please, say to me
You'll let me be your man
And please, say to me
You'll let me hold your hand
You'll let me hold your hand
I wanna hold your hand

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I say that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I feel that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand



Does the letter "z" ever appear in this song's lyrics? 

In [236]:
"z" in lyrics


False

Apparently not!

<span class="your_turn">Try it yourself: make a string</span>

* Find the lyrics to a song you like (or a speech, poem, etc.)
* Copy-paste the text into the Python notebook.
* Use triple-quotes `"""  """` to make a <span class="new_term">string</span> out of this text, and assign it to a variable.
* Does your song contain every letter of the alphabet? Use the `in` keyword to check for the least common letters (z, q, j, x).
<br><br><br>

<h2 class="fragment semi-fade-out"> Strings (continued) </h2>

In [239]:
print(lyrics.replace("hand", "elbow")) 



Oh yeah, I'll tell you something
I think you'll understand
When I say that something
I wanna hold your elbow
I wanna hold your elbow
I wanna hold your elbow

Oh please, say to me
You'll let me be your man
And please, say to me
You'll let me hold your elbow
You'll let me hold your elbow
I wanna hold your elbow

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I say that something
I wanna hold your elbow
I wanna hold your elbow
I wanna hold your elbow

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I feel that something
I wanna hold your elbow
I wanna hold your elbow
I wanna hold your elbow
I wanna hold your elbow



<br><br><br>

## Comparisons

There are several important symbols for comparing two numbers in Python.

In [154]:
6 > 5 # greater than

True

In [155]:
5 > 6 

False

In [156]:
5 < 5 # less than

False

In [158]:
5 <= 5 # less than or equal to (≤)

True

In [157]:
4 >= 5 # greater than or equal to (≥)

False

## Comparisons

Because the `=` character is used for assignment, we test whether two numbers are ***equal*** using `==`.

In [159]:
len("Timon") == len("Simba")

True

To test whether two numbers are unequal ≠, use `!=`.

In [160]:
len("Timon") != len("Pumbaa")

True

## Logic

The `and` keyword between two `True`/`False` <span class="new_term">conditionals</span> gives us 
* `True` if both are `True`
* `False` if either one or both is `False`. 

In [225]:
("a" in "cat") and ("g" in "dog") # True and True -> True

True

In [231]:
("a" in "cat") and ("h" in "dog") # True and False -> False

False

Similarly, the `or` keyword between two conditionals gives us 

* `True` if either one or both is `True`
* `False` if both are `False`. 

In [227]:
("a" in "cat") or ("h" in "dog") # True or False -> True

True

In [228]:
("a" in "cat") or ("d" in "dog") # True or True -> True

True

In [165]:
("h" in "cat") or ("k" in "dog") # False or False -> False 

False

`not` in front of any conditional turns `True` into `False` and vice versa.

In [166]:
not True

False

In [232]:
not "w" in "pig"

True

In [233]:
"w" not in "pig" # another way to write the same thing

True

<span class="your_turn">Try it yourself: Conditionals</span>

* Using your text from the previous activity, come up with two words that *are* in your text and two other words that *aren't* in your text.
* Make four conditionals of the format `<word> in <text>`. In my example, one conditional could be:

In [240]:
"wanna" in lyrics


True

* Use `and` and `or` to put conditionals together. For example,

In [242]:
("wanna" in lyrics) and ("pony" in lyrics)


False

* Try to come up with two such statements that evaluate to `True`, and two others that evaluate to `False`.

## Summary of Part I 

* Today we introduced Python's fundamentals for
    * math: <span class="new_term">integers</span> and <span class="new_term">floats</span>
    * text: <span class="new_term">strings</span>
    * comparisons: (`>`, `==`, etc.)
    * logic: `and`/`or`/`not`
* Tomorrow we will introduce <span class="new_term">functions</span>, which are the most important concept for making computer programs useful. <br><br><br>

<h1 class="r-stretch" style="text-align:center"> Introduction to Python, Part II  </h1>

<center>
    <br>
     <table style="background-color: white;">
        <tr>
            <td>
    <img src="https://www.python.org/static/community_logos/python-logo.png" width=250 style="float:center">
            </td>
            <td>
    <img src="https://brand.jhu.edu/assets/uploads/sites/5/2014/06/university.logo_.small_.horizontal.blue_.jpg" width=250 style="float:center">
            </td>
            </tr>
    </table>
        <br>
    Johns Hopkins University  <br>
Krieger School of Arts and Sciences  <br>
Office of Undergraduate Research, Scholarly and Creative Activity (URSCA) <br>
May 24, 2022 <br>
    <br>
    Prof. Daniel Beller <br>
    JHU Department of Physics &amp; Astronomy
    </center>
 </section>

## `if` statements

We usually want the program to *do* something if our <span class="new_term">conditional</span> evaluates to `True`. The `if` statement is fundamental to Python logic.

In [171]:
x = 5
y = 8

if x < 6: 
    print("x is less than 6") 
    y = 12                    
    
print("Now y equals", y)  

x is less than 6
Now y equals 12


Notice formatting:

* The `if` statement ends in colon `:`
* Following this is an <span class="new_term">indented code block</span> containing any number of lines. (Use 'tab' key.)
* Delete the indent to leave the `if` block.

<span class="your_turn">Try it!</span> – What happens if we change the conditional to a `False` one? 
<br><br><br>

### `if`-`else` 

Often we want our program to do something *different* if the conditional is `False`. For this, we follow the `if` block with an `else` block.


In [243]:
x = 5

if x > 6:
    y = 12
else:
    y = 9
    
print("y equals", y)

y equals 9


### `if`-`elif`-`else`

When there are more than two possibilities to consider, we can put one or more `elif` blocks between the `if` block and the `else` block. Each block is only executed if the conditional above it was `False`.

In [173]:
x = 6

if x > 6:
    y = 12
elif x == 6: # another conditional!
    y = 11
else:
    y = 10
    
print(y)

11


## `while` loops

What do we get if we add up all the integers $1, 2, \dots 100$?

It would be annoying if we had to type out all 100 numbers!

Fortunately, programming languages allow <span class="new_term">loops</span> that run a block of code repeatedly.

In [174]:
total = 0
n = 1

while n <= 100:         # loop until n > 100
    total = total + n   # add n to total        
    print(n, total)
    n = n + 1           # add 1 to n     

1 1
2 3
3 6
4 10
5 15
6 21
7 28
8 36
9 45
10 55
11 66
12 78
13 91
14 105
15 120
16 136
17 153
18 171
19 190
20 210
21 231
22 253
23 276
24 300
25 325
26 351
27 378
28 406
29 435
30 465
31 496
32 528
33 561
34 595
35 630
36 666
37 703
38 741
39 780
40 820
41 861
42 903
43 946
44 990
45 1035
46 1081
47 1128
48 1176
49 1225
50 1275
51 1326
52 1378
53 1431
54 1485
55 1540
56 1596
57 1653
58 1711
59 1770
60 1830
61 1891
62 1953
63 2016
64 2080
65 2145
66 2211
67 2278
68 2346
69 2415
70 2485
71 2556
72 2628
73 2701
74 2775
75 2850
76 2926
77 3003
78 3081
79 3160
80 3240
81 3321
82 3403
83 3486
84 3570
85 3655
86 3741
87 3828
88 3916
89 4005
90 4095
91 4186
92 4278
93 4371
94 4465
95 4560
96 4656
97 4753
98 4851
99 4950
100 5050


The answer is $5050$. <br><br>






## Functions 

<span class="new_term">Functions</span> are the most important concept for *reusable* programs. 

A <span class="new_term">function</span> is a rule for mapping inputs to outputs. 



This is similar to mathematical functions like 

$$ f(x) = 2x+1 $$ 

Here the input is $x$ and the output is $2x+1$. 

For example, an input of $3$ gets mapped to an output of $7$.

The equivalent function in Python looks like this:

In [175]:
def f(x):
    return 2 * x + 1

Important features:
* `def` makes `f` a function 
* input (which we call an "<span class="new_term">argument</span>") `x` goes inside parentheses `()`
* output follows `return` (on same line)
* colon and indent just like `if` blocks

Now we can use the function for any input number!

In [176]:
f(3)

7

In [177]:
f(-201)

-401

<span class="your_turn">Try it!</span> – Functions

Write a function that calculates a 6% Maryland sales tax on any price. 

Then use it to find the sales tax for a book marked \$12.99.

In [245]:
def sales_tax(price):
    amount = price * 0.06
    answer = round(amount, 2)
    return answer 

sales_tax(12.99)

0.78

<br><br><br>

## Lists and `for` loops

<span class="new_term">Lists</span> are one of Python's most convenient features. Make a list by putting any collection of values inside `[]` and separated by commas `,`.

In [256]:
my_list = [3, -201, 11.1]

A `for` loop is similar to a `while` loop, except that a `for` loop iterates over the contents of your list and then stops automatically.

In [257]:
for number in my_list: # for each item in my_list, labeled as "number"...
    print(2 * number)  # multiply "number" by 2 and show me the result

6
-402
22.2


Notice that `number` is just a placeholder, like $x$ in algebra. We'd get the same result with:

In [258]:
for banana in my_list:
    print(2 * banana)

6
-402
22.2


We can combine <span class="new_term">functions</span> with `for` loops. For example, using our `f(x)` from above:

In [259]:
for number in my_list:
    print(f(number))

7
-401
23.2


<br><br><br>

### `range`

<span class="fragment">In a `for` loop, if we have a sequence of numbers like</span>

In [260]:
for i in [1, 2, 3, 4]:
    print(f(i))

3
5
7
9


we can use a `range` instead:

In [261]:
for i in range(1, 5): # start at 1, stop BEFORE 5
    print(f(i))

3
5
7
9


Adding up all the integers $1, 2, \dots, 100$ is easier this way:

In [262]:
total = 0

for i in range(1, 101):
    total = total + i
    
print(total)

5050


Actually, Python gives us an even better shortcut with its `sum` function:

In [263]:
sum(range(1,101))

5050

<br><br><br>

### Working with strings in `for` loops

How many times do the Beatles say the word "hand" in the lyrics to "I Want to Hold Your Hand"?

First we can use the `.split()` method to chop up a string into a <span class="new_term">list</span> of one-word strings.

In [264]:
words = lyrics.split()

print(words)

['Oh', 'yeah,', "I'll", 'tell', 'you', 'something', 'I', 'think', "you'll", 'understand', 'When', 'I', 'say', 'that', 'something', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'Oh', 'please,', 'say', 'to', 'me', "You'll", 'let', 'me', 'be', 'your', 'man', 'And', 'please,', 'say', 'to', 'me', "You'll", 'let', 'me', 'hold', 'your', 'hand', "You'll", 'let', 'me', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'And', 'when', 'I', 'touch', 'you', 'I', 'feel', 'happy', 'Inside', "It's", 'such', 'a', 'feeling', 'that', 'my', 'love', 'I', "can't", 'hide', 'I', "can't", 'hide', 'I', "can't", 'hide', 'Yeah,', 'you', 'got', 'that', 'something', 'I', 'think', "you'll", 'understand', 'When', 'I', 'say', 'that', 'something', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'And', 'when', 'I', 'touch', 'you', 'I', 'feel', 'happy', 'Inside', "It's", 'such

In [265]:
total = 0

for word in words:
    if word == "hand":  # string equality comparison 
        total = total + 1
        
print("The word 'hand' appears", total, "times in this song.")       

The word 'hand' appears 13 times in this song.


<br><br><br>

### List elements

What is the 24th word in the song? Use square brackets `[]` to access an element of a list.

In [188]:
words[23]

'your'

Why `23` and not `24`? Because **Python starts counting from zero, not one.**

In [189]:
words[0] # 1st word

'Oh'

In [190]:
words[1] # 2nd word

'yeah,'

How many words are in the song? 

In [149]:
len(words)

182

What are the first 10 words in the song? For this we can use a <span class="new_term">slice</span>

In [199]:
first_ten = words[0:10] # start at 1st word, stop before 11th word
print(first_ten)

['Oh', 'yeah,', "I'll", 'tell', 'you', 'something', 'I', 'think', "you'll", 'understand']


We can glue these words back together into a single string (with spaces in between the words) in this way:

In [200]:
" ".join(first_ten)

"Oh yeah, I'll tell you something I think you'll understand"

When does the word "something" first appear in the song?

In [198]:
words.index("something") 

5

It's word number 6! (Since index 0 is word number 1.)  <br><br><br>

## List comprehension 

You can construct lists from `for` loops in a single line. This is called a <span class="new_term">list comprehension</span>.

For example, to show the length of each word in the lyrics (in order):

In [12]:
word_lengths = [ len(word) for word in words ]
print(word_lengths)

NameError: name 'words' is not defined

How many characters does the longest word have? 

In [11]:
max(word_lengths)

NameError: name 'word_lengths' is not defined

<h3 class="fragment semi-fade-out">List comprehension </h3>

We can put an `if` statement inside a list comprehension to filter the list before it goes through the `for` loop. 

For example, to get all the words in `lyrics` that start with "y":

In [252]:
my_list = [ word for word in words if word.startswith("y") ]
print(my_list)

['yeah,', 'you', "you'll", 'your', 'your', 'your', 'your', 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'your']


This is a shortcut for the following `for` loop with `if` block inside it:

In [251]:
my_list = []

for word in words:
    if word.startswith("y"):       
        my_list.append(word)
        
print(my_list)        
        

['yeah,', 'you', "you'll", 'your', 'your', 'your', 'your', 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'your']


<br><br><br>

## Summary

* `if`-`elif`-`else` statements allow your program to make "decisions"
* `while` loops, `for` loops, and functions let you use the same block of code repeatedly
* lists make it easy to work with data such as a collection of numbers or long text.

### Next steps:

You can download Python from www.python.org and run your own programs on your computer. 

There are many free resources to help you learn Python at you own pace. 

* "Automate the Boring Stuff with Python" [automatetheboringstuff.com](https://automatetheboringstuff.com), a beginner's guide to Python scripting for everyday use
* Many other options at [wiki.python.org](https://wiki.python.org) -> "Beginners Guide" 

Contact me at d.a.beller@jhu.edu if you have questions. These slides will be posted to [bellerphysics.gitlab.io/Code.html](https://bellerphysics.gitlab.io/Code.html) — which is a webpage created using Python!


<br><br><br>

With just a few lines of Python, you can examine publicly available data from anywhere on the Internet. 

The example below goes beyond what we'll cover here, but you too can become a data expert with Python.

In [80]:
import pandas 

data = pandas.read_csv("https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv")
MD_data = data.loc[data["Province_State"] == "Maryland"]
MD_numvax = MD_data.set_index("Date").loc[:,
    ["People_Fully_Vaccinated",
    "People_Partially_Vaccinated"]
]
MD_numvax.dropna().plot(
    title="Number of people vaccinated against COVID-19 in Maryland (source: coronavirus.jhu.edu)",
    backend="plotly"
)